### PR TITLE
Add support for other backends in get_preferred_device

### PR DESCRIPTION
**Pull request by jeejakp12**
Currenlty get_preferred_device supports only cuda and cpu. Add support for other backends using backend config.

Fixes #ISSUE_NUMBER

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 1]
```diff
@@ -370,6 +371,17 @@ def _get_preferred_device(self) -> torch.device:
         Return the preferred device to be used when creating tensors for collectives.
         This method takes into account the associated process group
         """
+        backend = dist.get_backend(self._process_group)
+        backend_config = dist.BackendConfig(dist.get_backend(self._process_group))
```
Comment (awgu) [inline torch/distributed/_shard/sharded_tensor/api.py:375]: nit: move this into the `else` branch since it is not used for NCCL case?

Comment (awgu) [reply torch/distributed/_shard/sharded_tensor/api.py:375]: and can reuse `backend` and not have to call `dist.get_backend` again?

Comment (jeejakp12) [reply torch/distributed/_shard/sharded_tensor/api.py:375]: yes, i will fix it, i forgot to remove the old code as well.

Comment (jeejakp12) [reply torch/distributed/_shard/sharded_tensor/api.py]: @awgu fixed the review comment.

**Review by awgu — DISMISSED** [review]

**Comment by jeejakp12** [comment]
@awgu  i have fixed the CI issue. made minor change, can you please help in review

**Comment by jeejakp12** [comment]
@awgu  can you help review the change?

**Review by kwen2501 — APPROVED** [review]
LGTM

**Comment by jeejakp12** [comment]
@kwen2501 @awgu  can you please review and help in merging? not sure about the issue with merge failure

**Comment by jeejakp12** [comment]
@kwen2501 there is some issue in merging, can you please help review and merge the change?

**Pull request closed without merging**